### PR TITLE
Bugfix: Added node properly in from_causal_graph

### PR DESCRIPTION
**Pull request by matteo-causalens**
## Description and Motivation
This PR:
- Fixed a bug in `cai_causal_graph.time_series_causal_graph.TimeSeriesCausalGraph.from_causal_graph` for `cai_causal_graph.time_series_causal_graph.TimeSeriesCausalGraph` where the method was not adding the node correctly.
- Added `max_time_lag` and `min_time_lag` properties to `cai_causal_graph.time_series_causal_graph.TimeSeriesCausalGraph` to return the maximum and minimum lag of the graph, respectively. The property `maxlag` was removed to avoid confusion.

## Types of changes

- [ ] Bug fix (non-breaking change which fixes an issue)
- [ ] New feature (non-breaking change which adds functionality)
- [ ] Improvement (non-breaking change which improves functionality)
- [ ] Breaking change (fix or feature that would cause existing functionality to change)
- [ ] Refactor (moving code around, general refactoring improvements)
- [ ] Documentation (documentation)

## PR Checklist:

- [ ] I have reviewed my own PR? (review the PR as you are reviewing someone else's PR)
- [ ] I have implemented all requirements? (see JIRA, project documentation)
- [ ] I am affecting someone else's work? If so: I have added those people as reviewers?
- [ ] I have added comments to all the bits that are hard to follow
- [ ] At a bare minimum, I tested this manually
- [ ] I have added unit test(s)
- [ ] Is this a bugfix? If so have I added a regression test?
- [ ] Does this PR satisfy the [one PR, one concern](https://medium.com/@fagnerbrack/one-pull-request-one-concern-e84a27dfe9f1) rule?
- [ ] If needed, documentation has been added/updated?
- [ ] I have updated the appropriate changelog with a line for my changes
- [ ] If this PR breaks reproducibility, have I added a note in the changelog explaining the break in reproducibility
      and its extent?

### CODE REVIEW

File: cai_causal_graph/time_series_causal_graph.py [hunk 1]
```diff
@@ -47,13 +47,41 @@ def wrapper(self: TimeSeriesCausalGraph, *args, **kwargs) -> Any:
...
+        self._max_time_lag = None
         self._variables = None
         return function
 
     return wrapper
 
 
+def get_min_lag(ts_graph: TimeSeriesCausalGraph) -> Optional[int]:
+    """
+    Get the minimum lag from a list of lags.
```
Comment (steven-causalens) [inline cai_causal_graph/time_series_causal_graph.py:59]: Should be "from a time series causal graph." instead of "from a list of lags."

File: cai_causal_graph/time_series_causal_graph.py [hunk 2]
```diff
@@ -47,13 +47,41 @@ def wrapper(self: TimeSeriesCausalGraph, *args, **kwargs) -> Any:
...
+
+    :param ts_graph: The time series causal graph.
+    return: The minimum lag.
```
Comment (steven-causalens) [inline cai_causal_graph/time_series_causal_graph.py:62]: ":return:"

**Review by steven-causalens — APPROVED** [review]
Just some minor comments

File: cai_causal_graph/time_series_causal_graph.py [hunk 3]
```diff
@@ -1137,16 +1169,28 @@ def adjacency_matrices(self) -> Dict[int, numpy.ndarray]:
         return adjacency_matrices
 
     @property
-    def maxlag(self) -> Optional[int]:
```
Comment (mmiklitz) [inline cai_causal_graph/time_series_causal_graph.py:1140]: @matteo-causalens, this is a breaking change for the graph API, should we maybe keep the property for backward compatibility and route it to return the new max_time_lag?

Comment (matteo-causalens) [reply cai_causal_graph/time_series_causal_graph.py:1140]: I can do it but I am not using it anymore so I don't see it as a big issue for the user. But of course I can add it

Comment (matteo-causalens) [reply cai_causal_graph/time_series_causal_graph.py:1140]: @mmiklitz Dara team is not using it yet, @tadhgokeeffe are you using it?

Comment (matteo-causalens) [reply cai_causal_graph/time_series_causal_graph.py:1140]: For now I can leave it to replicate the past correct behaviour

**Review by andrew-causalens — CHANGES_REQUESTED** [review]
This is super confusing. The max lag will actually be the longest time forward while min lag is the longest time backwards. I think we need to change the naming a bit.

If we have X-3 -> X-2 -> X-1 -> X -> X+1 -> X+2, then max lag will be +2 and min lag will be -3. I would think max lag is -3 and max future is 2. Min lag is really -1 in this case. Just by name I would think:

```
max_lag = min(node.time_lag for node in nodes if node.time_lag < 0)
min_lag = max(node.time_lag for node in nodes if node.time_lag < 0)

max_future = max(node.time_lag for node in nodes if node.time_lag > 0)
min_future = min(node.time_lag for node in nodes if node.time_lag > 0)
```

**Comment by max-causalens** [comment]
> This is super confusing. The max lag will actually be the longest time forward while min lag is the longest time backwards. I think we need to change the naming a bit.
> 
> If we have X-3 -> X-2 -> X-1 -> X -> X+1 -> X+2, then max lag will be +2 and min lag will be -3. I would think max lag is -3 and max future is 2. Min lag is really -1 in this case. Just by name I would think:
> 
> ```
> max_lag = min(node.time_lag for node in nodes if node.time_lag < 0)
> min_lag = max(node.time_lag for node in nodes if node.time_lag < 0)
> 
> max_future = max(node.time_lag for node in nodes if node.time_lag > 0)
> min_future = min(node.time_lag for node in nodes if node.time_lag > 0)
> ```

Sorry I have not been following too deeply, but how does it make sense at all to have any notion of future?  Why would you ever want to do that?  IMO, it should just be min_lag and max_lag and those should be positive integers.

**Comment by andrew-causalens** [comment]
> > This is super confusing. The max lag will actually be the longest time forward while min lag is the longest time backwards. I think we need to change the naming a bit.
> > If we have X-3 -> X-2 -> X-1 -> X -> X+1 -> X+2, then max lag will be +2 and min lag will be -3. I would think max lag is -3 and max future is 2. Min lag is really -1 in this case. Just by name I would think:
> > ```
> > max_lag = min(node.time_lag for node in nodes if node.time_lag < 0)
> > min_lag = max(node.time_lag for node in nodes if node.time_lag < 0)
> > 
> > max_future = max(node.time_lag for node in nodes if node.time_lag > 0)
> > min_future = min(node.time_lag for node in nodes if node.time_lag > 0)
> > ```
> 
> Sorry I have not been following too deeply, but how does it make sense at all to have any notion of future? Why would you ever want to do that? IMO, it should just be min_lag and max_lag and those should be positive integers.

Yes I agree they should be positive integers. I think @matteo-causalens needs these properties to address the coerce time tiers question from Tadhg. However, the current implementation on this PR is too confusing. We need an idea of future in the `TimeSeriesCausalGraph` as we can extend forward and backwards in time. So future needs to exist. However, I agree on Max's point on always positive. I would define them as so:

```
max_lag = max(abs(node.time_lag) for node in nodes if node.time_lag < 0)
min_lag = min(abs(node.time_lag) for node in nodes if node.time_lag < 0)
```

Then you would get 3 for `max_lag` and 1 for `min_lag`.

**Comment by matteo-causalens** [comment]
> > > This is super confusing. The max lag will actually be the longest time forward while min lag is the longest time backwards. I think we need to change the naming a bit.
> > > If we have X-3 -> X-2 -> X-1 -> X -> X+1 -> X+2, then max lag will be +2 and min lag will be -3. I would think max lag is -3 and max future is 2. Min lag is really -1 in this case. Just by name I would think:
> > > ```
> > > max_lag = min(node.time_lag for node in nodes if node.time_lag < 0)
> > > min_lag = max(node.time_lag for node in nodes if node.time_lag < 0)
> > > 
> > > max_future = max(node.time_lag for node in nodes if node.time_lag > 0)
> > > min_future = min(node.time_lag for node in nodes if node.time_lag > 0)
> > > ```
> > 
> > 
> > Sorry I have not been following too deeply, but how does it make sense at all to have any notion of future? Why would you ever want to do that? IMO, it should just be min_lag and max_lag and those should be positive integers.
> 
> Yes I agree they should be positive integers. I think @matteo-causalens needs these properties to address the coerce time tiers question from Tadhg. However, the current implementation on this PR is too confusing. We need an idea of future in the `TimeSeriesCausalGraph` as we can extend forward and backwards in time. So future needs to exist. However, I agree on Max's point on always positive. I would define them as so:
> 
> ```
> max_lag = max(abs(node.time_lag) for node in nodes if node.time_lag < 0)
> min_lag = min(abs(node.time_lag) for node in nodes if node.time_lag < 0)
> ```
> 
> Then you would get 3 for `max_lag` and 1 for `min_lag`.

It is useful in other cases (not only Tadhg's request). I agree that min and max are confusing but this is due to the fact that we discussed with Marcin time ago regarding the property `time_lag` and are  the names should change and reflect more the idea of time. Because min and max in the current implementation are correct names actually because they are signed. If we want them to be always positive then we must have both past and future lags. But also having `max_lag`and `min_lag` are confusion both in terms of names and in terms of value if you want to refer them as past value. 
Moreover, I don't need the intermediate ones (i.e. min_lag and min_future) and I don't see them super useful but I can add them.

So what I would propose is max_past_lag and max_future_lag, both positive.

**Comment by matteo-causalens** [comment]
> > > > This is super confusing. The max lag will actually be the longest time forward while min lag is the longest time backwards. I think we need to change the naming a bit.
> > > > If we have X-3 -> X-2 -> X-1 -> X -> X+1 -> X+2, then max lag will be +2 and min lag will be -3. I would think max lag is -3 and max future is 2. Min lag is really -1 in this case. Just by name I would think:
> > > > ```
> > > > max_lag = min(node.time_lag for node in nodes if node.time_lag < 0)
> > > > min_lag = max(node.time_lag for node in nodes if node.time_lag < 0)
> > > > 
> > > > max_future = max(node.time_lag for node in nodes if node.time_lag > 0)
> > > > min_future = min(node.time_lag for node in nodes if node.time_lag > 0)
> > > > ```
> > > 
> > > 
> > > Sorry I have not been following too deeply, but how does it make sense at all to have any notion of future? Why would you ever want to do that? IMO, it should just be min_lag and max_lag and those should be positive integers.
> > 
> > 
> > Yes I agree they should be positive integers. I think @matteo-causalens needs these properties to address the coerce time tiers question from Tadhg. However, the current implementation on this PR is too confusing. We need an idea of future in the `TimeSeriesCausalGraph` as we can extend forward and backwards in time. So future needs to exist. However, I agree on Max's point on always positive. I would define them as so:
> > ```
> > max_lag = max(abs(node.time_lag) for node in nodes if node.time_lag < 0)
> > min_lag = min(abs(node.time_lag) for node in nodes if node.time_lag < 0)
> > ```
> > 
> > 
> >     
> >       
> >     
> > 
> >       
> >     
> > 
> >     
> >   
> > Then you would get 3 for `max_lag` and 1 for `min_lag`.
> 
> It is useful in other cases (not only Tadhg's request). I agree that min and max are confusing but this is due to the fact that we discussed with Marcin time ago regarding the property `time_lag` and are the names should change and reflect more the idea of time. Because min and max in the current implementation are correct names actually because they are signed. If we want them to be always positive then we must have both past and future lags. But also having `max_lag`and `min_lag` are confusion both in terms of names and in terms of value if you want to refer them as past value. Moreover, I don't need the intermediate ones (i.e. min_lag and min_future) and I don't see them super useful but I can add them.
> 
> So what I would propose is max_past_lag and max_future_lag, both positive.

As I already mentioned to Andrew, I would like to be aware of the sign lags sometimes. For example, the graph X(t-3) -> X(t-2) would have -3 and -2 in my implementation. Instead if we move to abs values we loose this information

**Comment by matteo-causalens** [comment]
> > > > > This is super confusing. The max lag will actually be the longest time forward while min lag is the longest time backwards. I think we need to change the naming a bit.
> > > > > If we have X-3 -> X-2 -> X-1 -> X -> X+1 -> X+2, then max lag will be +2 and min lag will be -3. I would think max lag is -3 and max future is 2. Min lag is really -1 in this case. Just by name I would think:
> > > > > ```
> > > > > max_lag = min(node.time_lag for node in nodes if node.time_lag < 0)
> > > > > min_lag = max(node.time_lag for node in nodes if node.time_lag < 0)
> > > > > 
> > > > > max_future = max(node.time_lag for node in nodes if node.time_lag > 0)
> > > > > min_future = min(node.time_lag for node in nodes if node.time_lag > 0)
> > > > > ```
> > > > 
> > > > 
> > > > Sorry I have not been following too deeply, but how does it make sense at all to have any notion of future? Why would you ever want to do that? IMO, it should just be min_lag and max_lag and those should be positive integers.
> > > 
> > > 
> > > Yes I agree they should be positive integers. I think @matteo-causalens needs these properties to address the coerce time tiers question from Tadhg. However, the current implementation on this PR is too confusing. We need an idea of future in the `TimeSeriesCausalGraph` as we can extend forward and backwards in time. So future needs to exist. However, I agree on Max's point on always positive. I would define them as so:
> > > ```
> > > max_lag = max(abs(node.time_lag) for node in nodes if node.time_lag < 0)
> > > min_lag = min(abs(node.time_lag) for node in nodes if node.time_lag < 0)
> > > ```
> > > 
> > > 
> > >     
> > >       
> > >     
> > > 
> > >       
> > >     
> > > 
> > >     
> > >   
> > > Then you would get 3 for `max_lag` and 1 for `min_lag`.
> > 
> > 
> > It is useful in other cases (not only Tadhg's request). I agree that min and max are confusing but this is due to the fact that we discussed with Marcin time ago regarding the property `time_lag` and are the names should change and reflect more the idea of time. Because min and max in the current implementation are correct names actually because they are signed. If we want them to be always positive then we must have both past and future lags. But also having `max_lag`and `min_lag` are confusion both in terms of names and in terms of value if you want to refer them as past value. Moreover, I don't need the intermediate ones (i.e. min_lag and min_future) and I don't see them super useful but I can add them.
> > So what I would propose is max_past_lag and max_future_lag, both positive.
> 
> As I already mentioned to Andrew, I would like to be aware of the sign lags sometimes. For example, the graph X(t-3) -> X(t-2) would have -3 and -2 in my implementation. Instead if we move to abs values we lose this information

For extending the graph it's not an issue as we compute the minimal graph first (X(t-1) -> X(t)) and then we do the rest. So for now we can leave it as it is but I think we should revisit this in the future together with the time_lag name as @mmiklitz already suggested.

File: cai_causal_graph/time_series_causal_graph.py [hunk 4]
```diff
@@ -47,13 +47,40 @@ def wrapper(self: TimeSeriesCausalGraph, *args, **kwargs) -> Any:
...
+    Get the minimum lag from a TimeSeriesCausalGraph.
+
+    :param ts_graph: The time series causal graph.
+    :return: The minimum lag.
+    """
+    assert isinstance(
+        ts_graph, TimeSeriesCausalGraph
+    ), f'Expected ts_graph to be a TimeSeriesGraph, got type {type(ts_graph)}.'
+    nodes = ts_graph.get_nodes()
+    return min([node.time_lag for node in nodes]) if len(nodes) > 0 else None
```
Comment (andrew-causalens) [inline cai_causal_graph/time_series_causal_graph.py:67]: Minor point but this builds entire list first. `min(node.time_lag for node in nodes)` doesn't require constructing the full list.

**Review by andrew-causalens — COMMENTED** [review]
Sorry still have a few comments just to make it clearer.

File: cai_causal_graph/time_series_causal_graph.py [hunk 5]
```diff
@@ -47,13 +47,40 @@ def wrapper(self: TimeSeriesCausalGraph, *args, **kwargs) -> Any:
...
         self._summary_graph = None
         self._stationary_graph = None
-        self._maxlag = None
         self._variables = None
         return function
 
     return wrapper
 
 
+def get_min_lag(ts_graph: TimeSeriesCausalGraph) -> Optional[int]:
```
[the remark this answers is not in the document]
Comment (andrew-causalens) [reply cai_causal_graph/time_series_causal_graph.py:56]: But the names were not updated to reflect the property ones so it was confusing still what a max lag and min lag means.

File: cai_causal_graph/time_series_causal_graph.py [hunk 6]
```diff
@@ -1138,15 +1169,31 @@ def adjacency_matrices(self) -> Dict[int, numpy.ndarray]:
 
     @property
     def maxlag(self) -> Optional[int]:
+        """Return the maximum absolute past lag of the graph."""
+        # get the maximum lag of the nodes in the graph
+        min_lag = get_min_lag(self)
+        return abs(min_lag) if min_lag is not None else None
```
[the remark this answers is not in the document]
Comment (matteo-causalens) [reply cai_causal_graph/time_series_causal_graph.py:1175]: Wait, why `max_backward_lag = self.max_backward_lag`. I need to compute it every time because it could have changed.

[the remark this answers is not in the document]
Comment (matteo-causalens) [reply cai_causal_graph/time_series_causal_graph.py:1175]: Sorry I got confused with the properties. We were referring to different ones. Now it's clear.

File: changelog.md [hunk 7]
```diff
@@ -1,5 +1,16 @@
 # Changelog
 
+## NEXT
+
+- Fixed a bug in `cai_causal_graph.time_series_causal_graph.TimeSeriesCausalGraph.from_causal_graph` for
+  `cai_causal_graph.time_series_causal_graph.TimeSeriesCausalGraph` where the method was not adding floating nodes
+  correctly.
+- Added `cai_causal_graph.time_series_causal_graph.TimeSeriesCausalGraph.max_backward_lag` and
+  `cai_causal_graph.time_series_causal_graph.TimeSeriesCausalGraph.min_forward_lag` properties to
```
Comment (andrew-causalens) [inline changelog.md:9]: `max_forward_lag`

File: changelog.md [hunk 8]
```diff
@@ -1,5 +1,16 @@
...
+  `cai_causal_graph.time_series_causal_graph.TimeSeriesCausalGraph` to return the absolute backward and forward time
```
Comment (andrew-causalens) [inline changelog.md:10]: " ... return the absolute maximum backward and forward ..." - just add the work maximum

File: cai_causal_graph/time_series_causal_graph.py [hunk 9]
```diff
@@ -441,7 +441,7 @@ def extend_graph(
             # Start from 1 as 0 is already defined.
             # We cannot start directly from maxlag as it may be possible that not all the nodes from 1 to -maxlag are
             # defined (as they were not needed in the minimal graph).
-            maxlag = minimal_graph.maxlag
+            maxlag = minimal_graph.max_backward_lag if minimal_graph.max_backward_lag is not None else None
```
Comment (andrew-causalens) [inline cai_causal_graph/time_series_causal_graph.py:444]: This can just be `maxlag = minimal_graph.max_backward_lag` as if None you set to None anyways.
